Using a post-transform for the AST

choldgraf opened the pull request:
This PR implements the logic that myst-nb uses for handling outputs. It roughly does the following:

* Add a few more cell node types (CellInputNode, CellOutputNode, and CellOutputBundleNode)
* Wraps cells in these node types
* Execute will now add a CellOutputBundleNode with the outputs inserted into it
* a post-transform will now transform these into CellOutputNode, followed by another imagecollector step.


## Todo

Need to figure out how to get the testing working. Currently I believe it assumes that the code is executed at parse time, as opposed to in a post-transform. Anybody have an idea for how to do this? My sphinx+pytest skills are not great

akhmerov commented on jupyter_sphinx/__init__.py:
```diff
@@ -170,15 +179,28 @@ def setup(app):
-        man=render_container,
+        CellOutputBundleNode,
+        override=True,
+        html=(skip, None),
```
Would it make sense to add a `raise` transform that would raise a sphinx extension error, for catching unexpected cases when the transform didn't run?

choldgraf replied:
I'm not sure what a raise translator is, but my intuition is that this would halt the build if a builder that wasn't supported was used? I feel like that'd be extreme no?

akhmerov replied:
This would be to clearly signal, that the node isn't meant to appear in the finalized AST. Seems to make sense and might provide an early notification when something goes awry.

choldgraf replied:
Ahh good point, OK I get it. Would a `raise` translator be something built-in to sphinx? or something we define?

akhmerov replied:
I think it'd be similar to `skip`, which we define in the same file:

```python
def halt(self, node):
    raise sphinx.errors.ExtensionError('Unexpected node encountered ...')
```

I don't expect that sphinx itself would define this because we should be raising an extension error, I think.

choldgraf replied:
ok, I think the latest push adds this, let me know if that's not what you had in mind.

What are your thoughts about testing?

akhmerov replied:
Thanks, this looks good.

I think raising an error when that node is attached is largely an internal consistency check, more akin to an assertion. Therefore I think it is appropriate that it is not covered by tests.

chrisjsewell commented:
> Need to figure out how to get the testing working. Currently I believe it assumes that the code is executed at parse time, as opposed to in a post-transform. Anybody have an idea for how to do this? My sphinx+pytest skills are not great

See https://github.com/ExecutableBookProject/MyST-Parser/tree/master/tests/test_sphinx for how I test against *minimal* sphinx projects

choldgraf commented:
@chrisjsewell do you think that a cleaner way to test this, rather than relying on a bunch of one-off assert statements, would be to set up regression tests against the doctree generated by Sphinx? 

And if so, I guess that's what this is doing https://github.com/ExecutableBookProject/MyST-Parser/blob/e37dd0058cba23827b717a4ad0e902c096b8a94f/tests/test_sphinx/conftest.py#L92 ?

Is the idea that you first generate the "correct" XML, and then after that when the tests are run they just compare against that?

chrisjsewell commented:
> Is the idea that you first generate the "correct" XML, and then after that when the tests are run they just compare against that?

Yes that's correct [pytest-regressions](https://pytest-regressions.readthedocs.io/en/latest/) generates an output the first time you run `pytest`, then tests against that output in all subsequent runs (it can be regenerated with `pytest force-regen`)

Note that the stored doctree XML is the one saved after transforms, but **before** post-transforms. That's why it would also be necessary to test against the created HTML: https://github.com/ExecutableBookProject/MyST-Parser/blob/e37dd0058cba23827b717a4ad0e902c096b8a94f/tests/test_sphinx/conftest.py#L84 (unless there is a way to recover the doctree after post-transforms)

akhmerov commented on jupyter_sphinx/css/jupyter-sphinx.css:
```diff
@@ -57,7 +57,7 @@ div.jupyter_container {
   background-color: #f7f7f7; /* for haiku */
 }
-div.jupyter_container {
+div.jupyter_container * {
```
Is this change accidental? I think it reverts a recent CSS bugfix.

choldgraf replied:
probably yeah, nice catch

akhmerov commented on jupyter_sphinx/ast.py:
```diff
@@ -358,3 +400,50 @@ def get_widgets(notebook):
+            col.process_doc(self.app, node)
+
+
+def _return_first_node_type(node, node_type):
```
Perhaps `find_only_child_by_type`?

At the same time I'm wondering if using the unpacking idiom could be better than introducing this function. e.g.
`output_bundle_node, = cell_node.traverse(CellOutputBundleNode)`.

What do you think?

akhmerov commented:
About the testing strategy: @choldgraf isn't this a separate issue?

akhmerov commented:
@choldgraf do you understand why the tests fail?

akhmerov commented on tests/test_execute.py:
```diff
@@ -38,7 +38,7 @@ def doctree(source, config=None, return_warnings=False, entrypoint="jupyter_sphi
         app.build()
 
-        doctree = app.env.get_doctree("contents")
+        doctree = app.env.get_and_resolve_doctree("contents", app.builder)
```
@choldgraf following a sphinx code dive I figured that this should also apply the post-transforms.

@chrisjsewell does this make sense?

choldgraf replied:
do you mean calling this function also applies post-transforms?

akhmerov replied:
Yep. I verified that after this modification of the fixture, the produces doctrees are correct.

choldgraf replied:
huh, interesting...so, does that mean we need to change something or that it's OK as-is?

akhmerov replied:
The failures that I see now are due to the AST change (there's now `CellInputNode` containing a literal block and not just a literal block), and not due to the post-transform not being applied.

akhmerov replied:
...so unless I'm overlooking something, this should be good.

akhmerov commented:
The tests are still failing, but now the that's just because the AST changed and tests require an update.

akhmerov commented on jupyter_sphinx/ast.py:
```diff
@@ -358,3 +400,50 @@ def get_widgets(notebook):
+        # Image collect extra nodes from cell outputs that we need to process
+        for node in self.document.traverse(image):
+            # If the image node has `candidates` then it's already been processed
+            # as in-line markdown, so skip it
```
Not sure what you mean here with "inline markdown". Can you please explain?

choldgraf replied:
TBH I am not sure, I will try and spend a bit of time early this week to look through this again

akhmerov replied:
From the processing around I thought you meant that the image what in the document source already. It's just the "inline markdown" that got me confused.

choldgraf replied:
yeah I think I probably just used "markdown" because it was coming from myst-markdown, but the overall principle applies, I'll remove the use of "markdown"

akhmerov commented on jupyter_sphinx/ast.py:
```diff
@@ -358,3 +400,50 @@ def get_widgets(notebook):
+            # as in-line markdown, so skip it
+            if "candidates" in node:
+                continue
+            col = ImageCollector()
```
I'm a bit concerned that we might be not respecting the app settings here; could we reuse the image collector from the app somehow? I imagine different builders will have different image collectors.

choldgraf replied:
hmm that is a good question. It hasn't been an issue in the Jupyter Book project so far (and we've built both PDF as well as HTML outputs), though I agree it'd be better if we could reuse the image collector that the app uses

akhmerov replied:
I'm also OK leaving this as is and waiting for someone to report a problem.

Please add an extra comment why this is necessary though.

akhmerov commented on jupyter_sphinx/__init__.py:
```diff
@@ -34,6 +39,11 @@
 
+# Used for nodes that should be gone by rendering time (OutputMimeBundleNode)
+def halt(self, node):
+    raise ExtensionError((f"Rendering encountered a node type that should "
```
We list python 3.5 as the minimal requirement, so introducing f-strings also requires bumping the python version in setup and in travis config.

I personally don't mind bumping python to 3.6, but @choldgraf you know better the situation with its support across the board.

CC also @SylvainCorlay in case he has objectsions.

choldgraf replied:
I'm fine removing the f-string because I agree that deprecating 3.5 is a different issue than this PR addresses. My general strategy is to support the latest-3 versions of Python, unless there is a *really* good reason not to

choldgraf commented:
Just wanna give a heartbeat here - I didn't forget about this PR, just making a push on a "beta" release for Jupyter Book so most of my dev time is directed towards that

akhmerov commented:
Sure thing, looking forward to the beta release :rocket:

choldgraf commented:
ok just rebased and pushed to get the new testing infra working on this PR

choldgraf commented:
ok I think that I've addressed all your comments, but now I remember why this one had slowed down - almost all of the tests are now failing because the structure of the doctrees are hard-coded in there...I can take a stab at converting them over but I hadn't had the time to do it yet

akhmerov commented:
Before also the post transform didn't run, now it's only the ast that changed.

choldgraf commented on tests/test_execute/test_basic.xml:
```diff
@@ -0,0 +1,5 @@
+<JupyterCellNode classes="jupyter_cell" code_below="False" emphasize_lines="" hide_code="False" hide_output="False" linenos="False" linenostart="True" raises="True" stderr="False">
```
here's an example of the file that pytest-regressions uses to decide whether it should fail or not. in subsequent runs, if running the test results in a *different* structure than this XML file, then the test will fail

there is one file created per call to `file_regression.check()`

seanpue replied:
This looks very cool. I wonder though if it would be easier if the XML were included in the test file itself (read as text stream) rather than in separate files. What do you think?

choldgraf commented on tests/test_execute.py:
```diff
@@ -58,36 +58,26 @@ def doctree(
-    assert cell.attributes["linenos"] is False
-    assert cell.children[0].rawsource.strip() == "2 + 2"
-    assert cell.children[1].rawsource.strip() == "4"
+    file_regression.check(cell.pformat(), extension=".xml")
```
here is how we use `file_regression`. we pass it some data, and it checks the content of that data against a file that is written to disk in the repository. If they don't match, then the test will fail

choldgraf commented:
@akhmerov take a look at the latest push, it tests out `pytest-regressions` on two tests. I am curious if you think this would be a good pattern to use for our tests here...that way we can test the full XML structure of the `CellNode` objects, rather than picking out individual pieces from them to test manually.

See a couple github comments I made for what is happening under the hood

chrisjsewell commented:
Just a reminder; in this PR or another we also want to move this out of myst-nb: https://github.com/executablebooks/MyST-NB/blob/d72ebe707d9ca307b4479c07e3eeed0deba6b996/myst_nb/transform.py#L68

choldgraf commented:
@chrisjsewell ah right...is this just an inline function for the same thing, basically? And if so could we re-use the same base function for both?

akhmerov commented:
Re: pytest-regressions,

Yes, appears to be more practical than the current approach of describing the same data in Python.

choldgraf commented:
note - after thinking about it, I think we should start off by just making sure tests pass here. Then in another PR, we can do a pytest-regressions switch to make sure that the tests already pass as expected

chrisjsewell commented:
> .is this just an inline function for the same thing, basically?

yep, that we use for the inline glue functionality

> And if so could we re-use the same base function for both?

Quite possiblly, whatever works best 😁

choldgraf commented:
ok, the tests are now fixed on my end, let's see if they pass. If so, then I'm +1 on a merge and quick release. Then we can finish up the inline function (new issue here: https://github.com/jupyter/jupyter-sphinx/issues/123) and the pytest-regressions move (#122) in another PR to avoid delaying this one further.

akhmerov commented on jupyter_sphinx/ast.py:
```diff
@@ -127,20 +129,24 @@ def run(self):
+            emphasize_lines=hl_lines,
+            raises=self.options.get("raises"),
+            stderr=("stderr" in self.options),
+            classes=["jupyter_cell"],
```
This class is introduced by the PR, right? That seems like it would warrant documenting [here](https://jupyter-sphinx.readthedocs.io/en/latest/#styling-options).

akhmerov commented on jupyter_sphinx/ast.py:
```diff
@@ -127,20 +129,24 @@ def run(self):
+        )
+
+        # Add the input section of the cell, we'll add output at execution time
+        cell_input = CellInputNode(classes=["cell_input"])
```
Same remark about classes.

akhmerov commented on jupyter_sphinx/ast.py:
```diff
@@ -270,6 +303,9 @@ def cell_output_to_nodes(cell, data_priority, write_stderr, dir, thebe_config):
             data = output["data"][mime_type]
             if mime_type.startswith("image"):
+                ####################################
+                # TODO: Figure out how to handle either inline or absolute image paths
```
Which paths are you referring to over here? I thought that we're in full control of where we're writing the image here.

choldgraf replied:
I think this is a carry over from a previous bug? It works fine now in tests and such

akhmerov commented on jupyter_sphinx/ast.py:
```diff
@@ -373,3 +414,52 @@ def get_widgets(notebook):
+        thebe_config = self.config.jupyter_sphinx_thebelab_config
+
+        for cell_node in self.document.traverse(JupyterCellNode):
+            output_bundle_node = _find_only_child_by_type(
+                cell_node, CellOutputBundleNode
+            )
```
I propose to replace the function call with
```suggestion
            output_bundle_node, = cell_node.traverse(CellOutputBundleNode)
```
and similarly in the other two uses of the function `_find_only_child_by_type`.

The unpacking idiom is uncommon, and therefore has a chance to be confusing. But at the same time this implementation saves checking what the function actually does, and to an expert reader not familiar with the code base will be more understandable.

akhmerov commented:
@choldgraf aside from the remaining comments, everything looks good.

choldgraf commented:
OK, I think I addressed all your comments, WDYT @akhmerov ?

choldgraf commented:
(also, once you merge I am happy to try bumping the version to see if our new auto-deploy works)

akhmerov commented:
Thanks, all looks good to merge.

0.2.4 sounds good: all the changes should be backwards compatible, and there's useful new functionality.
